[ConstraintSystem] Handle ambiguities caused by requirement failures

xedin opened the pull request:
Aggregate all requirement failures (regardless of kind) that belong
to the same locator and diagnose them as an ambiguity (if there is
more than one overload) or as the singular failure if all solutions
point to the same overload.

Resolves: https://github.com/apple/swift/issues/56173
Resolves: rdar://79357256
Resolves: rdar://104135510

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSFix] NFC: Add a common base class for all requirement failures

xedin committed:
[CSFix] Move `diagnoseForAmbiguity` to RequirementFix

The logic is common for all types of fixes that represent
a requirement failure.

xedin committed:
[ConstraintSystem] Handle ambiguities caused by requirement failures

Aggregate all requirement failures (regardless of kind) that belong
to the same locator and diagnose them as an ambiguity (if there is
more than one overload) or as the singular failure if all solutions
point to the same overload.

LucianoPAlmeida commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -4427,6 +4427,71 @@ static bool diagnoseAmbiguityWithContextualType(
+
+    if (overloadChoices.size() > 0) {
+      // Some of the choices are non-declaration,
+      // let's deligate that to ambiguity diagnostics.
```
Typo deligate -> delegate?

LucianoPAlmeida replied:
Nevermind, I just didn't know [deligate](https://www.thefreedictionary.com/Deligate) was a word. Although the context delegate also makes sense :)

xedin replied:
I will switch it, no problem :) I’d like to make some changes to address the FIXME anyway

xedin replied:
✅

LucianoPAlmeida commented:
@xedin Looking at the logic, I wonder if that could be also fixing https://github.com/apple/swift/issues/56173? It was some time ago but I do remember seeing requirement fixes, so maybe is possible ...

xedin commented:
@LucianoPAlmeida it might indeed fix that, I add a test-case.

xedin commented:
@LucianoPAlmeida Case from [this comment](https://github.com/apple/swift/issues/56173#issuecomment-1108434102) is indeed improved by these changes, the rest are diagnosed property as well.

xedin committed:
[ConstraintSystem] Handle requirement failures associated with injected result builder methods

Only way to aggregate build* methods is via source locations because
each result builder kind would create a new build* expression node.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please clean test Linux platform

LucianoPAlmeida commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -4850,14 +4911,83 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
+            // Ambiguity only if all of the solutions have a requirement
+            // fix at the given location.
+            if (aggregate.size() == solutions.size())
+              viableGroups.push_back(std::move(aggregate));
```
Since aggregate is `const` I think move here results in a copy anyways.  Overload resolution would end picking up const ref overload of `push_back` if I understand correctly.

xedin replied:
Hm I thought that `SmallVector` has a `const SmallVector<T> &&` overload for constructor but apparently not, although maybe it still makes sense to keep it to show intent because the vector passed there is taken from the map...

LucianoPAlmeida replied:
Does `aggregate` need to be `const`? If is just a non-const ref then it can be moved. 
```cpp
#include <iostream>
#include <vector>
class A {
public:
  A() {}
  A(A&&a) noexcept { std::cout << "Move" << std::endl;  }
  A(const A&a) noexcept { std::cout << "Copy" << std::endl; }
};

int main(int argc, const char * argv[]) {
  std::vector<A> v;
  auto l = [&](const A&a) {
    v.push_back(std::move(a));
  };
  auto l1 = [&](A&a) {
    v.push_back(std::move(a));
  };
  
  A a;
  l(a); // Copy
  l1(a); // Move
  return 0;
}

```

xedin replied:
Let me try, I though that we kind of stuck with `const` here, but maybe not.

xedin replied:
Looks like we can indeed drop `const`!

LucianoPAlmeida replied:
Cool :)

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

hborla reviewed (approved):
Nice!

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
@swift-ci please test macOS platform

LucianoPAlmeida commented:
@swift-ci please test macOS platform

compnerd commented:
Please test with following PRs:
apple/swift-driver#1281

@swift-ci please smoke test macOS platform

xedin commented:
Please test with following PRs:
https://github.com/apple/swift-driver/pull/1281

@swift-ci please smoke test macOS platform

compnerd commented:
@swift-ci please smoke test macOS platform

xedin commented:
full suite is failing with unrelated failures:

```
Failed Tests (6):
  Swift(watchsimulator-x86_64) :: Interpreter/SDK/check_class_for_archiving_log.swift
  Swift(watchsimulator-x86_64) :: stdlib/NSSlowString.swift
  Swift(watchsimulator-x86_64) :: stdlib/StringIndex.swift
  Swift-validation(watchsimulator-x86_64) :: stdlib/String.swift
  Swift-validation(watchsimulator-x86_64) :: stdlib/StringBreadcrumbs.swift
  Swift-validation(watchsimulator-x86_64) :: stdlib/StringUTF8.swift
```

```
xcrun: error: sh -c '/Applications/Xcode-beta.app/Contents/Developer/usr/bin/xcodebuild -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -find dyld_info 2> /dev/null' failed with exit code 17664: (null) (errno=Invalid argument)
xcrun: error: unable to find utility "dyld_info", not a developer tool or in PATH
falling back to 'xcrun dyldinfo' ...
Child process terminated with signal 6: Abort trap
```
